Round clock testing

kolmanv opened the pull request.

gshilin commented on email.gemspec:
```diff
@@ -0,0 +1,13 @@
```
Why don't you use mailer from Rails?

kolmanv replied:
Why not use the simple Mail gem? ...As Action Mailer now uses the Mail gem...
There is an option to use Log4r as mailing feature (maybe wrong approach, but seems for testing server is ok).

gshilin commented on lib/content_data/content_data.rb:
```diff
@@ -483,16 +484,15 @@ def deep_check(instance)
 
-    # @raise [ArgumentError] when instance_check_level is incorrect
     def check_instance(instance)
       case Params['instance_check_level']
```
func = "#{Params['instance_check_level']}_check"
send(func, instance) rescue throw ArgumentError.new "Unsupported check level #{Params['instance_check_level']}"

gshilin commented on lib/content_data/content_data.rb:
```diff
@@ -553,6 +553,166 @@ def get_query(variable, params)
     end
 
-    private :shallow_check, :deep_check, :check_instance, :get_query
+    private :shallow_check, :deep_check, :check_instance
```
Is it your code convention to supply a list of functions?
I've never seen it before. Just "private" and afterwards all private functions' definitions.

gshilin commented on lib/content_data/content_data.rb:
```diff
@@ -553,6 +553,166 @@ def get_query(variable, params)
+  # @raise [ArgumentError] when instance_check_level is incorrect
+  def validate(params = nil)
+    # used to answer whether specific param was set
+    param_exists = Proc.new do |param|
```
Add "params" as parameter and define this function outside of #validate.
There is no need to do it again and again...

gshilin commented on lib/content_data/content_data.rb:
```diff
@@ -553,6 +553,166 @@ def get_query(variable, params)
+    process_params = Proc.new do |values|
+      # values is a Hash with keys: :content, :instance and value appropriate to key
+      if param_exists.call :failed
+        unless values[:content].nil?
```
I would do this and next "if" to be one-liners.
params[:failed].add_content values[:content] unless values[:content].nil?
params[:failed].add_instance values[:instance] unless values[:instance].nil?

Short and clear.

gshilin commented on lib/content_data/content_data.rb:
```diff
@@ -553,6 +553,166 @@ def get_query(variable, params)
+      unless check_instance instance
+        is_valid = false
+
+        unless params.nil? || params.empty?
```
Just add this condition to #process_params, the code here will be cleaner

gshilin commented on lib/content_data/content_data.rb:
```diff
@@ -553,6 +553,166 @@ def get_query(variable, params)
+  end
+
+  # @raise [ArgumentError] when instance_check_level is incorrect
+  def check_instance(instance)
```
I've seen this function already on line 487 and ever wrote a comment :)

gshilin commented on lib/content_data/content_data.rb:
```diff
@@ -553,6 +553,166 @@ def get_query(variable, params)
+      is_match = false
+      var_value = instance.instance_variable_get("@#{variable}")
+
+      if exact.include? var_value
```
is_match = (exact.include? var_value) || (from.nil? || var_value > from) && (to.nil? || var_value < to)
and there is no need to initialize it separately tree lines above.

gshilin commented on lib/content_data/content_data.rb:
```diff
@@ -553,6 +553,166 @@ def get_query(variable, params)
+    end
+
+    result_index = ContentData.new
+    instances.each_value do |instance|
```
Why don't you use inject? That's exactly what it does: gets an initial value (ContentData.new in this case) and returns it after all calculations.

gshilin commented on lib/email/email.rb:
```diff
@@ -0,0 +1,87 @@
+require 'net/smtp'
```
Why don't you use https://github.com/mikel/mail ? That's standard client, tested by millions of users, support gmail (see its wiki page).
There is no need to re-invent a wheel :-1:

gshilin commented on lib/file_utils.rb:
```diff
@@ -1,3 +1,4 @@
 require 'file_utils/file_utils'
+
 module FileUtils
```
Why do you need this definition here? Don't you have it in the required file?

gshilin commented on lib/email.rb:
```diff
@@ -0,0 +1,4 @@
+require 'email/email'
```
Why do you need this definition here? Don't you have it in the required file?

gshilin commented on lib/file_utils/file_generator/file_generator.rb:
```diff
@@ -6,22 +6,25 @@
+require 'log'
+
 module FileGenerator
-  Params.string('target_path', '~/.bbfs/test_files', 'Represents the target path for files generation')
```
Once I've asked you to check http://ruby-doc.org/stdlib-1.9.3/libdoc/ostruct/rdoc/OpenStruct.html

yarondbb committed:
Added testing server locally on same machine

yarondbb committed:
Added testing server locally on same machine

yarondbb committed:
removed file as part of log infrastructure change to log4r

yarondbb committed:
Updated file as part of log infrastructure change to log4r

yarondbb committed:
Masked off the logging part. Need to use the new log4r

yarondbb committed:
Added testing server locally on same machine

yarondbb committed:
Added log of backup diff opearion

yarondbb committed:
Updated log

yarondbb committed:
Updated log

yarondbb committed:
Updated log

yarondbb committed:
updated log version

yarondbb committed:
Merge remote-tracking branch 'origin/round_clock_testing' into round_clock_testing

Conflicts:
	lib/content_server/remote_content.rb

yarondbb committed:
1. Fixed file init. File is deleted if eists. Dir created unless exists

2. Enabled email.
       Each error or fatal will be flushed imidiately to email.
       Added Log.flush method to allow manual flush of all outputters (including email) by the user.

genadyp committed:
general exception handling

yarondbb committed:
Fixed flush code so that flush will not send empty emails

yarondbb committed:
Removed password from code

Added some basic documentation

genadyp committed:
gmail password update

yarondbb committed:
1.Added new log param to enable flush at each message (default is flush)
2. Updated logs yml examples
3. Added support to flush log outputters to console and file if program terminates (in bin files)

kolmanv committed:
Update log.rb

Added log debug level.

kolmanv committed:
Update testing_server.rb

Remove from_email and from_email_password as they are defined in log.

kolmanv committed:
Update testing_server.rb

Parameter already defined in log.

yarondbb committed:
removed redundent log tests (using tested em log4r)
Updated log. If init is not executed then init will be called on first attemp to log (same behavior as old log)

yarondbb committed:
fixed tests

kolmanv committed:
Update backup_server

Fix small retries calculations.

kolmanv committed:
Update content_server

Fix small retries calculation in content server
